File: examples/src/main/java/com/teamdev/jxbrowser/examples/SelectClientCertificate.java
/*
 *  Copyright 2023, TeamDev. All rights reserved.
 *
 *  Redistribution and use in source and/or binary forms, with or without
 *  modification, must retain the above copyright notice and the following
 *  disclaimer.
 *
 *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
 *  "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
 *  LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
 *  A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
 *  OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
 *  SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
 *  LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
 *  DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
 *  THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 *  (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 *  OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

package com.teamdev.jxbrowser.examples;

import static com.teamdev.jxbrowser.engine.RenderingMode.HARDWARE_ACCELERATED;
import static javax.swing.JOptionPane.PLAIN_MESSAGE;

import com.teamdev.jxbrowser.browser.Browser;
import com.teamdev.jxbrowser.browser.callback.SelectClientCertificateCallback;
import com.teamdev.jxbrowser.engine.Engine;
import com.teamdev.jxbrowser.view.swing.BrowserView;
import java.awt.BorderLayout;
import java.awt.event.WindowAdapter;
import java.awt.event.WindowEvent;
import java.security.cert.X509Certificate;
import java.util.List;
import java.util.Objects;
import java.util.stream.IntStream;
import javax.swing.JFrame;
import javax.swing.JOptionPane;
import javax.swing.SwingUtilities;
import javax.swing.WindowConstants;

/**
 * This example demonstrates how to display "Select Client SSL Certificate"
 * dialog where an end-user must select an SSL certificate to continue loading
 * the web page.
 *
 * <p>Important: before you run this example, please follow the instruction at
 * <a href="https://badssl.com/download/">badssl.com</a> and install
 * the required custom SSL certificate.
 */
public final class SelectClientCertificate {

    private static BrowserView view;

    public static void main(String[] args) {
        Engine engine = Engine.newInstance(HARDWARE_ACCELERATED);
        Browser browser = engine.newBrowser();

        SwingUtilities.invokeLater(() -> {
            view = BrowserView.newInstance(browser);

            JFrame frame = new JFrame("Select Client SSL Certificate");
            frame.addWindowListener(new WindowAdapter() {
                @Override
                public void windowClosing(WindowEvent e) {
                    engine.close();
                }
            });
            frame.setDefaultCloseOperation(WindowConstants.EXIT_ON_CLOSE);
            frame.add(view, BorderLayout.CENTER);
            frame.setSize(700, 500);
            frame.setLocationRelativeTo(null);
            frame.setVisible(true);
        });

        browser.set(SelectClientCertificateCallback.class, (params, tell) -> {
            List<X509Certificate> certificates = params.certificates().stream()
                    .map(cerf -> cerf.toX509Certificate().orElse(null))
                    .filter(Objects::nonNull)
                    .toList();

            if (certificates.isEmpty()) {
                tell.cancel();
            }

            Object[] selectionValues = IntStream.range(0, certificates.size())
                    .mapToObj(i -> new SelectionValue(i, certificates.get(i)))
                    .toArray();

            Object selectedValue = JOptionPane.showInputDialog(
                    view,
                    params.message(), params.title(),
                    PLAIN_MESSAGE, null,
                    selectionValues, selectionValues[0]
            );

            if (selectedValue != null) {
                SelectionValue selected = (SelectionValue) selectedValue;
                tell.select(selected.index);
            }
        });

        browser.navigation().loadUrl("https://client.badssl.com/");
    }

    private record SelectionValue(int index, X509Certificate certificate) {

        @Override
        public String toString() {
            return certificate.getSubjectX500Principal().getName();
        }
    }
}

File: examples/src/main/kotlin/com/teamdev/jxbrowser/examples/SelectClientCertificate.kt
/*
 *  Copyright 2024, TeamDev. All rights reserved.
 *
 *  Redistribution and use in source and/or binary forms, with or without
 *  modification, must retain the above copyright notice and the following
 *  disclaimer.
 *
 *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
 *  "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
 *  LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
 *  A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
 *  OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
 *  SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
 *  LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
 *  DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
 *  THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 *  (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 *  OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

package com.teamdev.jxbrowser.examples

import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.selection.selectable
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.Card
import androidx.compose.material.Text
import androidx.compose.runtime.*
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.unit.dp
import androidx.compose.ui.window.Dialog
import androidx.compose.ui.window.DialogProperties
import androidx.compose.ui.window.singleWindowApplication
import com.teamdev.jxbrowser.browser.callback.SelectClientCertificateCallback
import com.teamdev.jxbrowser.compose.BrowserView
import com.teamdev.jxbrowser.dsl.Engine
import com.teamdev.jxbrowser.dsl.browser.navigation
import com.teamdev.jxbrowser.dsl.register
import com.teamdev.jxbrowser.engine.RenderingMode
import com.teamdev.jxbrowser.net.tls.ClientCertificate
import com.teamdev.jxbrowser.net.tls.SslPrivateKey
import java.security.cert.X509Certificate
import kotlin.jvm.optionals.getOrNull

/**
 * This example demonstrates how to display "Select Client SSL Certificate"
 * dialog where you must select a required SSL certificate to continue loading
 * web page.
 *
 * **Important:** before you run this example, please follow the instruction
 * at [badssl.com](https://badssl.com/download/) and install the required
 * custom SSL certificate.
 */
fun main() = singleWindowApplication(title = "Select client SSL certificate") {
    val engine = remember { Engine(RenderingMode.OFF_SCREEN) }
    val browser = remember { engine.newBrowser() }

    var x509Certificates by remember { mutableStateOf(emptyList<X509Certificate>()) }
    var onCertificateSelected by remember { mutableStateOf<(X509Certificate) -> Unit>({ }) }

    BrowserView(browser)

    if (x509Certificates.isNotEmpty()) {
        SelectionDialog(
            certificates = x509Certificates,
            onSelected = onCertificateSelected
        )
    }

    LaunchedEffect(Unit) {
        browser.register(SelectClientCertificateCallback { params, tell ->
            val certificates = params.certificates()
            if (certificates.isEmpty()) {
                tell.cancel()
            }

            x509Certificates = certificates
                .map { it.toX509Certificate() }
                .mapNotNull { it.getOrNull() }

            onCertificateSelected = { selectedCertificate ->

                x509Certificates = emptyList()

                // TODO:2024-04-10:yevhenii.nadtochii: Not working.
                //  See issue: https://github.com/TeamDev-IP/JxBrowser-Docs/issues/932
                val privateKey = SslPrivateKey.of(selectedCertificate.encoded)
                val clientCertificate = ClientCertificate.of(selectedCertificate, privateKey)

                // Pass the selected certificate to the engine.
                tell.select(clientCertificate)
            }
        })

        browser.navigation.loadUrl("https://client.badssl.com/")
    }
}

@Composable
private fun SelectionDialog(
    certificates: List<X509Certificate>,
    onSelected: (X509Certificate) -> Unit,
) = Dialog(
    onDismissRequest = { /* Selection is mandatory. */ },
    properties = DialogProperties(usePlatformDefaultWidth = false)
) {
    Card(
        modifier = Modifier.size(600.dp, 300.dp),
        shape = RoundedCornerShape(16.dp),
    ) {
        Column(
            modifier = Modifier.padding(16.dp),
            verticalArrangement = Arrangement.spacedBy(16.dp),
        ) {
            Text("Select a certificate: ")
            SelectionList(certificates, onSelected)
        }
    }
}

@Composable
private fun SelectionList(
    certificates: List<X509Certificate>,
    onSelected: (X509Certificate) -> Unit
) {
    var selected by remember { mutableStateOf<X509Certificate?>(null) }
    Column {
        certificates.forEach { certificate ->
            Text(
                text = certificate.subjectX500Principal.name,
                modifier = Modifier
                    .selectable(
                        selected = certificate == selected,
                        onClick = {
                            selected = certificate
                            onSelected(certificate)
                        }
                    )
                    .background(
                        if (selected == certificate) Color.Gray
                        else Color.Transparent
                    )
                    .padding(8.dp)
            )
        }
    }
}
